src/coverlet.core/Coverage.cs

```diff
@@ -147,55 +147,52 @@ private void CalculateCoverage()
-                    using (var gz = new GZipStream(fs, CompressionMode.Decompress))
-                    using (var sr = new StreamReader(gz))
+                    // File not instrumented, or nothing in it called.  Warn about this?
+                    continue;
+                }
+
+                using (var fs = new FileStream(result.HitsFilePath, FileMode.Open))
+                using (var sr = new StreamReader(fs))
+                {
+                    string row;
+                    while ((row = sr.ReadLine()) != null)
                     {
-                        string row;
-                        while ((row = sr.ReadLine()) != null)
-                        {
-                            var info = row.Split(',');
-                            // Ignore malformed lines
-                            if (info.Length != 4)
-                                continue;
+                        var info = row.Split(',');
+                        // Ignore malformed lines
+                        if (info.Length != 5)
+                            continue;
 
-                            bool isBranch = info[0] == "B";
+                        bool isBranch = info[0] == "B";
 
-                            var document = result.Documents.FirstOrDefault(d => d.Path == info[1]);
-                            if (document == null)
-                                continue;
+                        if (!result.Documents.TryGetValue(info[1], out var document))
+                        {
+                            continue;
+                        }
 
-                            int start = int.Parse(info[2]);
+                        int start = int.Parse(info[2]);
+                        int hits = int.Parse(info[4]);
 
-                            if (isBranch)
-                            {
-                                uint ordinal = uint.Parse(info[3]);
-                                var branch = document.Branches.First(b => b.Number == start && b.Ordinal == ordinal);
-                                if (branch.Hits != int.MaxValue)
-                                    branch.Hits += branch.Hits + 1;
-                            }
-                            else
+                        if (isBranch)
+                        {
+                            int ordinal = int.Parse(info[3]);
+                            var branch = document.Branches[(start, ordinal)];
+                            branch.Hits = hits;
+                        }
+                        else
+                        {
+                            int end = int.Parse(info[3]);
+                            for (int j = start; j <= end; j++)
                             {
-                                int end = int.Parse(info[3]);
-                                for (int j = start; j <= end; j++)
-                                {
-                                    var line = document.Lines.First(l => l.Number == j);
```

Comment:
This line was the core of the bottleneck (together with the similar line doing `document.Branches.First` above.  Having the lines in a list means that it finding the correct line for an event will have to search through `n/2` items of the `n` items in the list on average. The number of events in a test can be described as `kn`, where `k` is the number of times the average line is executed during the test.

Putting that together means that processing all events in the test will take `(k/2) * n^2` operations. In complexity theory the factor is not interesting since when `n` gets big, the value of `k` is irrelevant, so the time complexity of this operation is expressed as `O(n^2)` in the big-O notation, i.e. quadratic time complexity.